lib/linter/linter.js

```diff
@@ -1324,6 +1691,46 @@ class Linter {
         return this._verifyWithoutProcessors(textOrSourceCode, config, options);
     }
 
+    /**
+     * Verify a given code with a flat config.
+     * @param {string|SourceCode} textOrSourceCode The source code.
+     * @param {FlatConfigArray} configArray The config array.
+     * @param {VerifyOptions&ProcessorOptions} options The options.
+     * @returns {LintMessage[]} The found problems.
+     */
+    _verifyWithFlatConfigArray(textOrSourceCode, configArray, options) {
+        debug("With flat config: %s", options.filename);
+
+        // we need a filename to match configs against
+        const filename = options.filename || "<input>";
+
+        // Store the config array in order to get plugin envs and rules later.
+        internalSlotsMap.get(this).lastConfigArray = configArray;
+        const config = configArray.getConfig(filename);
+
+        // Verify.
+        if (config.processor) {
+            debug("Apply the processor: %o", config.processor);
+            const { preprocess, postprocess, supportsAutofix } = config.processor;
+            const disableFixes = options.disableFixes || !supportsAutofix;
+
+            return this._verifyWithFlatConfigArrayAndProcessor(
+                textOrSourceCode,
+                config,
+                { ...options, filename, disableFixes, postprocess, preprocess },
+                configArray
+            );
+        }
+
+
+        // check for options-based processing
+        if (options.preprocess || options.postprocess) {
+            return this._verifyWithFlatConfigArrayAndProcessor(textOrSourceCode, config, options);
+        }
```

Comment:
In the `if (config.processor)` path, `_verifyWithFlatConfigArray` calls `_verifyWithFlatConfigArrayAndProcessor` with `options.preprocess`, and then `_verifyWithFlatConfigArrayAndProcessor` calls `_verifyWithFlatConfigArray` with the same `options.preprocess`, so this will run `config.processor.preprocess()` again on its output:

```js
const { Linter } = require("eslint");

const linter = new Linter({ configType: "flat" });

const config = {
    files: ["*.md"],
    processor: {
        preprocess(text, filename) {
            console.log(`Called with text = '${text}', filename = '${filename}'`);

            return [{ text: "bar", filename: "0.js" }];
        },
        postprocess() {
            return [];
        }
    }
};

linter.verify("foo", config, "a.md");

// logs:
//       Called with text = 'foo', filename = 'a.md'
//       Called with text = 'bar', filename = 'a.md/0_0.js'
```

Reply:
Huh, nice catch. I'll look at that.

Just FYI: It was a little bit difficult to understand that you were pointing out a problem here. When you state the behavior you're seeing, it would help me out a lot if you'd say something like, "This behavior is unexpected and I don't see it when using eslintrc." This is probably just the brain fog, but little hints like saying "unexpected" make it a lot easier for me to decipher descriptions.